pkg/sql/lease.go

```diff
@@ -785,22 +820,56 @@ func (t *tableState) acquireNodeLease(
 		return errors.New("cannot acquire lease when draining")
 	}
 
-	// Notify when lease has been acquired.
+	// Notify when lease is being acquired.
 	t.mu.acquiring = make(chan struct{})
+	t.mu.latestAcquisitionErr = nil
+
+	err := t.stopper.RunAsyncTask(
+		context.Background(), "sql.tableState: async table lease acquisition", func(ctx context.Context) {
+			err := t.acquireNodeLeaseBlocking(ctx, m, minExpirationTime)
+			if err != nil {
+				log.Error(ctx, errors.Wrapf(err, "Async TableDescriptor lease acquisition error"))
+			}
+		})
+
+	if err != nil {
+		close(t.mu.acquiring)
+		t.mu.acquiring = nil
+	}
+
+	return err
+}
+
+// acquireNodeLeaseBlocking does the blocking operation for acquiring a
+// table descriptor lease.
+//
+// t.mu.acquiring must be initialized before calling.
+func (t *tableState) acquireNodeLeaseBlocking(
+	ctx context.Context, m *LeaseManager, minExpirationTime hlc.Timestamp,
+) error {
+	t.mu.Lock()
+	defer t.mu.Unlock()
 	defer func() {
 		close(t.mu.acquiring)
 		t.mu.acquiring = nil
 	}()
-	// We're called with mu locked, but need to unlock it during lease
-	// acquisition.
+
+	if m.isDraining() {
```

Comment:
you've already checked this. no need to check it again

Reply:
I'm a bit skeptical of this because if the system starts draining by the time this routine begins.

Sidenote: Is draining what happens when cockroach is cleaning quitting? Can we possibly run into an error state if we don't stop here while draining is true?